Codec/Archive/Tar/Check.hs

```diff
@@ -65,24 +65,46 @@ checkSecurity = checkEntries checkEntrySecurity
 checkEntrySecurity :: Entry -> Maybe FileNameError
 checkEntrySecurity entry = case entryContent entry of
     HardLink     link -> check (entryPath entry)
-                 `mplus` check (fromLinkTarget link)
+                 `mplus` checkLink (entryPath entry) (fromLinkTarget link)
     SymbolicLink link -> check (entryPath entry)
-                 `mplus` check (fromLinkTarget link)
+                 `mplus` checkLink (entryPath entry) (fromLinkTarget link)
     _                 -> check (entryPath entry)
 
   where
+
+    checkCommon name =
+        FilePath.Native.isAbsolute name || not (FilePath.Native.isValid name)
+
     check name
-      | FilePath.Native.isAbsolute name
+      | checkCommon name
       = Just $ AbsoluteFileName name
 
-      | not (FilePath.Native.isValid name)
+      | any (=="..") (FilePath.Native.splitDirectories name)
       = Just $ InvalidFileName name
 
-      | any (=="..") (FilePath.Native.splitDirectories name)
+      | otherwise = Nothing
+
+    -- checkLink introduced to handle https://github.com/haskell/tar/issues/32
+    checkLink name link
+      | checkCommon name
+      = Just $ AbsoluteFileName name
+
+      | linkDepth link name > 0
       = Just $ InvalidFileName name
 
       | otherwise = Nothing
 
+linkDepth :: FilePath -- ^ Name of link
+          -> FilePath -- ^ Contents of link
+          -> Int
+linkDepth link name =
+  let allPaths = FilePath.Native.splitDirectories link ++ FilePath.Native.splitDirectories name
+  in getDepth allPaths
+
+  where getDepth []         = 0
+        getDepth ("..":fps) = 1 + getDepth fps
+        getDepth (_:fps)    = getDepth fps - 1
```

Comment:
I feel uneasy merging such changes without any regression tests. Could we please add some?

Reply:
What happens if one of the elements is `"."`? It should not affect the depth, right? Imagine a symbolic link of form `./.././..` or similar. 

@hasufell is my understanding correct here?

Reply:
Yeah, I think we need to `normalise` first. I followed the codepath and I don't see that normalisation happens anywhere else.